tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala

```diff
@@ -0,0 +1,135 @@
+package scala.scalanative
+package compiler
+package pass
+
+import nir._, Shows._
+import util.sh
+
+object GlobalValueNumbering {
+
+  type Hash = Int
+
+  class HashFunction(hashLocal: Local => Hash) extends (Any => Hash) {
+
+    /* Careful : the hashes are not independent !
+     * ex: %src.1 = stackalloc[ty]
+     * %src.2 = load[ty]: %src.1: ptr
+     * then H(%src.2) = H("Load") XOR H(%src.1) XOR H(ty)
+     *                = H("Load") XOR H("Stackalloc") XOR H(ty) XOR H(ty)
+     *                = H("Load") XOR H("Stackalloc")
+     *                  -> independent of the type
+     * This might not appear as a problem (the example given is not hashed like this anymore)
+     * If it is a problem, we can still hash the XOR, i.e. :
+     * %src.1 = H(H("Stackalloc") XOR H(ty))
+     * But this increases the cost and requires a more powerful hashing function for integers than H(n) = n
+     * */
+
+    import HashFunction._
+
+    def apply(obj: Any): Hash = {
```

Comment:
It was only to simplify the code for `hashOp`, but it is not strictly necessary. I am still unsure whether this or "specializing" `hashOp` is better. I would say the latter to be honest, so long and tedious code change ahead.